src/context.rs

```diff
@@ -10,6 +10,211 @@ use crate::ffi::types::{c_uint, c_void, AlignedType};
+        lock: &'a SpinLock,
+    }
+
+    impl Deref for SpinLockGuard<'_> {
+        type Target = ([AlignedType; MAX_PREALLOC_SIZE], Option<NonNull<ffi::Context>>);
+        fn deref(&self) -> &Self::Target {
+            // Safe: we hold the lock.
+            unsafe { &*self.lock.data.get() }
+        }
+    }
+
+    impl DerefMut for SpinLockGuard<'_> {
+        fn deref_mut(&mut self) -> &mut Self::Target {
+            // Safe: mutable access is unique while the guard lives.
+            unsafe { &mut *self.lock.data.get() }
+        }
+    }
+
+    impl Drop for SpinLockGuard<'_> {
+        fn drop(&mut self) { self.lock.unlock(); }
+    }
+
+    /// Borrows the global context and do some operation on it.
+    ///
+    /// If `randomize_seed` is provided, it is used to rerandomize the context after the
+    /// operation is complete. If it is not provided, randomization is skipped.
+    ///
+    /// Only a bit or two per signing operation is needed; if you have any entropy at all,
+    /// you should provide it, even if you can't provide 32 random bytes.
+    pub fn with_global_context<T, Ctx: Context, F: FnOnce(&Secp256k1<Ctx>) -> T>(
+        f: F,
+        rerandomize_seed: Option<&[u8; 32]>,
+    ) -> T {
+        with_raw_global_context(
+            |ctx| {
+                let secp = ManuallyDrop::new(Secp256k1 { ctx, phantom: PhantomData });
+                f(&*secp)
+            },
+            rerandomize_seed,
+        )
+    }
+
+    /// Borrows the global context as a raw pointer and do some operation on it.
+    ///
+    /// If `randomize_seed` is provided, it is used to rerandomize the context after the
+    /// operation is complete. If it is not provided, randomization is skipped.
+    ///
+    /// Only a bit or two per signing operation is needed; if you have any entropy at all,
+    /// you should provide it, even if you can't provide 32 random bytes.
+    pub fn with_raw_global_context<T, F: FnOnce(NonNull<ffi::Context>) -> T>(
+        f: F,
+        rerandomize_seed: Option<&[u8; 32]>,
+    ) -> T {
+        assert!(
+            unsafe {
+                ffi::secp256k1_context_preallocated_size(AllPreallocated::FLAGS)
+                    <= core::mem::size_of::<[AlignedType; MAX_PREALLOC_SIZE]>()
+            },
+            "prealloc size exceeds our guessed compile-time upper bound"
+        );
+
+        // Our function may be expensive, so before calling it, we copy the global
+        // context into this local buffer on the stack. Then we can release it,
+        // allowing other callers to use it simultaneously.
+        let mut store = [AlignedType::ZERO; MAX_PREALLOC_SIZE];
+        let buf = NonNull::new(store.as_mut_ptr() as *mut c_void).unwrap();
+
+        let ctx = match SECP256K1.try_lock() {
+            None => unsafe {
+                // If we can't get the lock, just do everything on the stack.
+                ffi::secp256k1_context_preallocated_create(buf, AllPreallocated::FLAGS)
```

Comment:
Ah, interesting, we can't use our high-level API because the preallocated types are different. But I think it could be sound to treat owned type as preallocated? Basically the same as you passing `&*boxed` to a function expecting `&T` if you have `Box<T>`.

Reply:
We can't use the high-level API because it's going away once we use the global context everywhere. So I just inlined it into this one place where it'd be used.

Reply:
LOL, true, that's was a huge brain fart. :D Though having an internal thin safe wrapper could be helpful at some places.